lib/server.js

```diff
@@ -68,64 +67,111 @@ module.exports = Server;
  * @param {string} [opts.username] Username for RabbitMQ.
  * @param {string} [opts.password] Username for Password.
  */
-function Server(opts) {
+function Server (opts) {
+  this._tasks = {};
   this.opts = opts || {};
-  defaults(this.opts, Server.DEFAULT_OPTIONS);
+  defaults(this.opts, {
+    hostname: process.env.RABBITMQ_HOSTNAME || 'localhost',
+    port: process.env.RABBITMQ_PORT || 5672,
+    username: process.env.RABBITMQ_USERNAME || 'guest',
+    password: process.env.RABBITMQ_PASSWORD || 'guest'
+  });
 
-  if (!Array.isArray(this.opts.queues)) {
-    throw new Error('ponos.Server: missing required `queues` option.');
-  }
+  if (this.opts.hermes) {
+    this.hermes = this.opts.hermes;
+  } else {
+    if (!Array.isArray(this.opts.queues)) {
+      throw new Error('ponos.Server: missing required `queues` option.');
+    }
+    if (!this.opts.queues.every(isString)) {
+      throw new Error(
+        'ponos.Server: each element of `queues` must be a string.'
+      );
+    }
 
-  if (!this.opts.queues.every(isString)) {
-    throw new Error('ponos.Server: each element of `queues` must be a string.');
+    this.hermes = hermes.hermesSingletonFactory(this.opts);
   }
 
-  if (this.opts.hermes) {
-    this.hermes = hermes;
-  }
-  else {
-    this.hermes = Promise.promisifyAll(
-      hermes.hermesSingletonFactory(this.opts)
-    );
-  }
+  this.hermes = Promise.promisifyAll(this.hermes);
 };
 
 /**
- * Default options for the worker server constructor.
- * @type {Object}
+ * Starts the worker server and listens for jobs coming from all queues.
+ * @return {promise} A promise that resolves when the server is listening.
  */
-Server.DEFAULT_OPTIONS = {
-  hostname: process.env.RABBITMQ_HOSTNAME || 'localhost',
-  port: process.env.RABBITMQ_PORT || 5672,
-  username: process.env.RABBITMQ_USERNAME || 'guest',
-  password: process.env.RABBITMQ_PASSWORD || 'guest'
+Server.prototype.start = function () {
+  return this.hermes.connectAsync().bind(this)
+    .then(this._assertHaveAllTasks)
+    .then(this._subscribeAll);
 };
 
 /**
- * Starts the worker server and listens for jobs coming from all queues.
- * @return {Promise} A promise that resolves when the server is listening.
+ * Helper function to ensure we have all required task handlers for the queue
+ * array with which the class was defined.
+ * @return {promise} Resolved when all tasks verified.
  */
-Server.prototype.start = function () {
-  var self = this;
-  return this.hermes.connectAsync()
-    .then(Promise.map(self.hermes.queues, self._subscribe));
+Server.prototype._assertHaveAllTasks = function () {
```

Comment:
This is needed at the moment because of how hermes works enforcing a given number of queues up front. In the future we may be able to dispense with this check.